server/src/main/java/org/elasticsearch/index/reindex/RejectAwareActionListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.reindex;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.CheckedConsumer;
+
+import java.util.function.Consumer;
+
+// public for testing
+public interface RejectAwareActionListener<T> extends ActionListener<T> {
+    void onRejection(Exception e);
+
+    /**
+     * Return a new listener that delegates failure/reject to original but forwards response to responseHandler
+     */
+    default <X> RejectAwareActionListener<X> withResponseHandler(Consumer<X> responseHandler) {
```

Comment:
My preference is for `default` to be used in a scenario where you are providing default behavior for an interface that some implementation *might* want to overwrite. This feels more like a helper static method such as:

```java
static <X> RejectAwareActionListener<X> withResponseHandler(RejectAwareActionListener<X> errorDelegate, Consumer<X> responseHandler) {
// Implementation
}
```

Reply:
I agree. Fixed in 07d479303e7b6add9cdb5deea506dafeaa3cd7ed